### DISCUSSION TITLE
Conveyors and force zones

### DISCUSSION THREAD

**Discussion by Zeblote** [NVIDIA-Omniverse/PhysX#280] [Q&A]
Hi, I'm wondering what would be the recommended ways to do these 3 things:

- While an object is touching another object, a constant directional force is added (it should act as a constraint between the two, imagine like a conveyor belt surface, that doubles as tank tracks when you put it on the ground)
- While an object is inside a potentially moving trigger, a constant directional force is added (it should act as a constraint between two actors, like above)
- While an object is inside a potentially moving trigger, a force is added accelerating the object towards a point (like a spherical gravity generator)

Are there already snippets doing something like this? I didn't immediately find them but maybe I didn't look hard enough.

**Comment by vreutskyy** [comment]
Hello @Zeblote,

For the 1st case, PhysX provides contact modification, which allows you to modify the contact properties during collision detection. You can use the `PxContactModifyCallback` to adjust the contact points and apply a constant directional force. Additionally, you can set a target velocity for the contact using the `PxContactSet::setTargetVelocity` method. This can simulate the effect of a conveyor belt or tank tracks. You can find more details here:
- [Contact Modification](https://nvidia-omniverse.github.io/PhysX/physx/5.3.1/docs/AdvancedCollisionDetection.html?highlight=modification#contact-modification)
- [Contact Target Velocity](https://nvidia-omniverse.github.io/PhysX/physx/5.3.1/_api_build/class_px_contact_set.html#_CPPv4N12PxContactSet17setTargetVelocityE5PxU32RK6PxVec3)

For the 2nd case, you can use trigger shapes to detect when an object enters a specific area. By setting up a trigger, you can apply a constant directional force to the object while it is inside the trigger volume. This can be done by modifying the contacts only when the trigger is hit. More information on triggers can be found here:
- [Trigger Shapes](https://nvidia-omniverse.github.io/PhysX/physx/5.3.1/docs/RigidBodyCollision.html?highlight=triggers#trigger-shapes)

For the 3rd case, you can use a similar trigger setup as in the 2nd case. When an object enters the trigger volume, you can apply a force that accelerates the object towards a specific point, simulating a spherical gravity generator. This can be achieved by applying forces to rigid bodies that hit the trigger. Detailed instructions on applying forces and torques are available here:
- [Applying Forces and Torques](https://nvidia-omniverse.github.io/PhysX/physx/5.3.1/docs/RigidBodyDynamics.html#applying-forces-and-torques)

I hope this helps!
